syntax = "proto3";

package users;

/*
* Messages used to handle requests
* to the user endpoint
*/

// Message used to store the user data 
// in the statefun and also what we give back to the
// queries
message UserData{
    uint64 id = 1;
    int64 credit = 2;
}

// Message used to keep track of the current max
// id to be assigned
message Count{
    uint64 num = 1;
}

// USER CREATION MESSAGES

message CreateUserRequest{
    string request_id = 1;
}


// internal to be used by the statefun
// between the create_user and the user function
message CreateUserWithId{
    uint64 id = 1;
    string request_id = 2;
}


// Message containing the others
message UserRequest{

    // GET USER MESSAGE
    message FindUserRequest{
    uint64 id = 1;
}

    // USER DELETION MESSAGES
    message RemoveUserRequest{
        uint64 id = 1;
    }

    // SUBTRACT CREDIT MESSAGES
    message SubtractCreditRequest{
        uint64 id = 1;
        int64 amount =2;
    }

    // ADD CREDIT MESSAGES
    message AddCreditRequest{
        uint64 id = 1;
        int64 amount = 2;
    }

    string request_id = 1;

    oneof message {
        FindUserRequest find_user = 2;
        RemoveUserRequest remove_user = 3;
        SubtractCreditRequest subtract_credit = 4;
        AddCreditRequest add_credit = 5;
    }

}

// Global response given for user operationsusers_pb2.py
message UserResponse{
    string request_id = 1;
    string result = 2;
}
